Bluetooth: add callbacks to indicate if pairing was complete or failed

jli157 opened the pull request:
Added callback functions for Bluetooth stack to notify its client that pairing has been completed or failed.

fixes: #8390

jli157 committed:
bluetooth: add callback to notify pairing is complete

Added two new callbacks for Bluetooth stack to notify
the application that pairing has been completed or failed.

fixes: #8390

Signed-off-by: Jun Li <jun.r.li@intel.com>

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -392,6 +392,8 @@ struct bt_conn_auth_cb {
 	void (*pincode_entry)(struct bt_conn *conn, bool highsec);
 #endif
+	/* @brief indicate the authentication process is complete */
+	void (*pairing_complete)(struct bt_conn *conn, unsigned int status);
```
Some issues I see with this:

 - The meaning and possible values for status will need to be documented. 

 - Are there defines/enum available for the possible values in some public header file? It seems this maps to the SMP status which is only available in the private `subsys/bluetooth/host/smp.h` header file currently.

 - Why `unsigned int` instead of `u8_t` as smp.c uses internally?

 - Since this maps to SMP errors it seems you don't plan to have this used for BR/EDR? Is it not useful there? Either way it needs to be documented that the application should not expect this callback for BR/EDR.

jli157 replied:
Hi Johan, I added a enumeration to address your concern. But maybe we need to make the naming better. Any suggestions?  

As for BR/EDR, do you mean the suggestion you made last time about moving the callback to "struct bt_conn_cb"? I'm not familiar with BR/EDR part, can you confirm and advise further on this?

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1472,6 +1472,10 @@ static void smp_pairing_complete(struct bt_smp *smp, u8_t status)
 	BT_DBG("status 0x%x", status);
 
+	if(bt_auth->pairing_complete) {
+		bt_auth->pairing_complete(smp->chan.chan.conn, status);
```
I think I suggested passing the no-bonding/bonding information to the callback, so the app has some idea whether the pairing will be discarded after the connection goes away, or if the keys will be stored persistently. Wouldn't this be useful to have?

jli157 replied:
Can you check the callback definition again? Also, for your convenience, the suggestions you made last time can be referred here: https://lists.zephyrproject.org/g/devel/topic/22117716#4763

jhedberg reviewed (changes requested):
Thanks a lot for taking the initiative and implementing a solution for this open issue! There are however some things to sort out before we can merge (see my inline comments).

Also, could you please make the commit message prefix consistent with the existing commit history. As you'll see we use "Bluetooth: ..." for all Bluetooth commits.

jli157 commented:
Hi @jhedberg , thanks for the comments! I just want to make this work to roll so that we can get the feature done as soon as possible. ( I need it. :) ). Please feel free to make more comments and directly modify it if necessary.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -1472,6 +1472,14 @@ static void smp_pairing_complete(struct bt_smp *smp, u8_t status)
 
+	if (bt_auth->pairing_completed) {
+		bt_pairing_result_t pairing_ret = status == 0 ?
+			BT_PAIRING_STATUS_BONDED :
```
I'm not sure you understand what bonded means. There's a flag in the smp struct for that. "bonded" means you're paired and intend to store the keys persistently. "paired" is a superset of "bonded" which doesn't guarantee that the keys will be stored persistently. In the Zephyr case when pairing, but not bonding, the keys are not stored beyond the connection that creates them, i.e. they're lost once you disconnect. So mapping "status == 0" to "bonded" is not right.

jli157 replied:
Thanks for clarifying! Where can the function "amp_pairing_complete" know that the pairing is "bonded" from its argument? The "bonded" status has been stored somewhere?

jli157 commented:
Reworked the patch by following your suggestions, and added two callbacks for notifying success and failure cases. 

However, for the failure reason, I need more advice about what kind of reasons should be used for the failure case.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -392,6 +392,23 @@ struct bt_conn_auth_cb {
+	 * @param conn Connection object.
+	 * @param err_code error code of failure.
+	 */
+	void (*pairing_failed)(struct bt_conn *conn, int err_code);
```
As long as we're not exposing the possible values in a public header, I'd question whether we should have an error parameter at all. Are there any good use cases where the app might want to do different things depending on the error?

jli157 replied:
Okay, maybe it is enough to keep just one callback like before, with a `ret_code` argument which is defined to serval values including:
```
BT_PAIRING_FAILED,
BT_PAIRING_NOT_BONDED,
BT_PAIRING_BONDED
```

jhedberg reviewed (dismissed):
This is starting to look much better, thanks. I'm still a bit concerned with the error code parameter (see my separate comment)

jhedberg commented:
This looks pretty good to me now. It'd be nice to hear some opinion from other maintainers though. @Vudentz @sjanc does this look reasonable, or do you see something obvious missing or something?

jhedberg commented:
> @Vudentz @sjanc does this look reasonable, or do you see something obvious missing or something?

I'll take the lack of comments from @sjanc and @Vudentz to mean that they're fine with this ;)
